lib/sycamore/sycamore/writers/duckdb_writer.py

```diff
@@ -0,0 +1,98 @@
+from dataclasses import dataclass, asdict
+from typing import Optional, Any
+from sycamore.writers.common import drop_types
+from typing_extensions import TypeGuard
+
+from sycamore.data.document import Document
+from sycamore.writers.base import BaseDBWriter
+import duckdb
+
+
+@dataclass
+class DuckDBClientParams(BaseDBWriter.ClientParams):
+    db_name: Optional[str] = None
+
+
+@dataclass
+class DuckDBTargetParams(BaseDBWriter.TargetParams):
+    table_name: str
+
+    def compatible_with(self, other: BaseDBWriter.TargetParams) -> bool:
+        if not isinstance(other, DuckDBTargetParams):
+            return False
+        if self.table_name != other.table_name:
+            return False
+        return True
+
+
+class DuckDBClient(BaseDBWriter.Client):
+    def __init__(self, client_params: DuckDBClientParams):
+        dict_params = asdict(client_params)
+        db_name = dict_params["db_name"] if "db_name" in dict_params else ":memory:"
+        self._client = duckdb.connect(db_name)
+
+    @classmethod
+    def from_client_params(cls, params: BaseDBWriter.ClientParams) -> "DuckDBClient":
+        assert isinstance(params, DuckDBClientParams)
+        return DuckDBClient(params)
+
+    def write_many_records(self, records: list[BaseDBWriter.Record], target_params: BaseDBWriter.TargetParams):
```

Comment:
How does the process model work here? Sycamore/Ray will spin up multiple processes, each of which can write out records. What will happen in this case -- will there be a separate duckdb instance in each process?

Reply:
I don't know what the right answer is here, but we should figure out what we want -- is it all processes writing to a single duckdb instance in one process?

Reply:
All processes now write to independent csv files that we then bulk load into a DuckDB database helping us get a lot of the benefits from multiprocessing as discussed